x-pack/elastic-agent/pkg/composable/vars.go

```diff
@@ -0,0 +1,186 @@
+
+// Replace returns a new value based on variable replacement.
+func (v *Vars) Replace(value string) (string, []map[string]interface{}, error) {
+	var processors []map[string]interface{}
+	c, err := ucfg.NewFrom(v.Mapping, ucfg.PathSep("."))
+	if err != nil {
+		return "", nil, err
+	}
+	if !validBrackets(value) {
+		return "", nil, fmt.Errorf("malformatted variable substitution")
+	}
+
+	result := ""
+	lastIndex := 0
+	for _, r := range varsRegex.FindAllSubmatchIndex([]byte(value), -1) {
+		for i := 0; i < len(r); i += 4 {
+			varContent := value[r[i+2]:r[i+3]]
+			invalidErr := fmt.Errorf("invalid variable: %s", value[r[i]:r[i+1]])
+			values, ok := splitPipes(varContent)
+			if !ok {
+				return "", nil, invalidErr
+			}
+			set := false
+			for _, val := range values {
+				if val == "" {
+					continue
+				}
+				if val[0] == '"' || val[0] == '\'' {
+					result += value[lastIndex:r[0]] + val[1:len(val)-1]
+					set = true
+					break
+				}
+				if val[len(val)-1] == '.' {
+					return "", nil, invalidErr
+				}
+				replace, err := c.String(val, -1, ucfg.PathSep("."))
+				if err == nil {
+					result += value[lastIndex:r[0]] + replace
+					set = true
+					if v.ProcessorsKey != "" && varPrefixMatched(val, v.ProcessorsKey) {
+						processors = v.Processors
+					}
+					break
+				}
+			}
+			if !set {
+				return "", nil, ErrNoMatch
+			}
+			lastIndex = r[1]
+		}
+	}
+	return result + value[lastIndex:], processors, nil
+}
+
+// validBrackets returns true when all starting {{ have a matching ending }}.
+func validBrackets(s string) bool {
+	result := ""
+	lastIndex := 0
+	match := false
+	for _, r := range varsRegex.FindAllSubmatchIndex([]byte(s), -1) {
+		match = true
+		for i := 0; i < len(r); i += 4 {
+			result += s[lastIndex:r[0]]
+			lastIndex = r[1]
+		}
+	}
+	if !match {
+		return !strings.Contains(s, "{{")
+	}
+	return !strings.Contains(result, "{{")
+}
+
+// stripSpace removes all the spaces unless they are quoted
+func stripSpace(s string) ([]rune, bool) {
+	const out = rune(0)
+
+	quote := out
+	escape := false
+	rs := make([]rune, 0, len(s))
+	for _, r := range s {
+		if !escape {
+			if r == '"' || r == '\'' {
+				if quote == out {
+					// start of unescaped quote
+					quote = r
+				} else if quote == r {
+					// end of unescaped quote
+					quote = out
+				}
+			}
+		}
+		// escape because of backslash (\); except when it is the second backslash of a pair
+		escape = !escape && r == '\\'
+		if quote != out || !unicode.IsSpace(r) {
+			rs = append(rs, r)
+		}
+	}
+	if quote != out {
+		return []rune(""), false
+	}
+	return rs, true
+}
+
+// stripEscapes strips escapes from quoted strings.
+func stripEscapes(s string) string {
+	if s == "" {
+		return s
+	}
+	if s[0] == '\'' {
+		return strings.Replace(s, `\'`, `'`, -1)
+	}
+	if s[0] == '"' {
+		return strings.Replace(s, `\"`, `"`, -1)
+	}
+	return s
+}
+
+// splitPipes split value at |, ignoring | in quotes.
+func splitPipes(input string) ([]string, bool) {
```

Comment:
Couldn't we use one of the existing handlebar libs for these kind of things?

Reply:
I wish I could have used a golang handlebars implementation instead of having to do it myself. But ultimately decided that it was best not to because:

1. Handlebars doesn't support have any notion of `|`. To perform fallbacks or anything like that when it comes to Handlebars you need to do `{{# if ...}` or `{{#unless ...}`.
2. Our implementation needed to know if a variable replace did occur and it came from a dynamic provider that the processors also needed to be attached to the `StrVar` from the `AST`. No Handlebars implementation exposes that information, it is all hidden in there implementation.
3. Our implementation also needed to know if a variable substitution did **not** match. Only then we are able to remove that input because the variable substitution did **not** occur. Again this is hidden in the Handlebars implementations.

Reply:
Makes sense. Your point 1 makes me think if handlebars is perhaps not the right pick as we seem to extend on it and this might be even confusing?

Reply:
We probably should stop calling it Handlerbars, and just say our variable replacement uses the same syntax as Handlerbars. But it is not Handlebars.